stub/src/main/java/io/grpc/stub/ServerCalls.java

```diff
@@ -169,6 +171,7 @@ public void onHalfClose() {
         }
 
         method.invoke(request, responseObserver);
+        request = null;
```

Comment:
It looks like you only need to add this line and not bother with the others. It doesn't seem `messageReceived` is improving any behavior.

Reply:
done